Binding model: Bindless part 2

doitsujin opened the pull request:
Implements initial support for bindless SRVs and Samplers, as well as dynamic and non-uniform descriptor indexing for Shader Model 5.1/6.0.

This consists of several parts, all of which depend on each other:
1. Shader compiler work to support these features, including some required additions to the interface between vkd3d and the shader compiler.
2. A complete rewrite of the root signature and resource binding code, which was needed since assumptions about the exact descriptor set layouts were duplicated in various places and the whole thing was barely maintainable, let alone extensible.
3. API-side additions to support bindless SRVs/Samplers.

The implementation passes both `test_bindless_srv_sm51` and `test_bindless_samplers_sm51` on AMDVLK as well as RADV with ACO (non-uniform indexing is broken on LLVM), and was further tested in Shadow of the Tomb Raider, Metro Exodus, and Resident Evil 2.

## Overview
Each descriptor heap holds one or more descriptor sets, one for each supported descriptor type. These sets are updated by functions such as `CreateShaderResourceView` and `CopyDescriptors`, and are bound to a command list via `SetDescriptorHeaps`.

The set and binding numbers for these sets are be the same for all pipeline layouts, so there is no need to rebind the sets when the root signature changes.

Descriptor table offsets are passed to the pipeline via push constants, one 32-bit integer per table, and are updated after each call to `Set*RootDescriptorTable`.

```glsl
/* Declarations */
layout(push_constant, std430) uniform push_cb {
    uint descriptor_tables[2];
} pc;
/* Descriptor heaps */
layout(set = 0, binding = 0) uniform sampler samplers[];
layout(set = 1, binding = 0) uniform usamplerBuffer srvBuf[];
layout(set = 2, binding = 0) uniform utexture2D srv2D[];
/* Root descriptor */
layout(set = 3, binding = 0, r32ui) uniform writeonly uimageBuffer u0;
/* Sample lookup */
texelFetch(srv2D[nonuniformEXT(pc.descriptor_tables[0u] + uint(floatBitsToInt(r0.x)))], ivec2(0), 0)
```

Root descriptors were also reworked to always use a separate set, even if `VK_KHR_push_descriptor` is not supported. This simplifies the design and may improve performance, since root descriptors likely change more frequently than descriptor tables. Static samplers also use a separate set now.

The packed descriptor set, root descriptor set and static sampler set are optional and will not be created if not needed by the root signature.

## Current issues
- It is not entirely clear which IDs we need to decorate with `NonUniformEXT` (is there any documentation for this?). RADV seems to be happy with pretty much anything, whereas AMDVLK requires it on both the index and the resource ID (of `OpTypeImage`/`OpTypeSampler`).
Decorating the pointer returned by `OpAccessChain` does *not* work, which complicates the design and causes some inevitable code duplication.
- We currently only enable bindless if the device reports `RESOURCE_BINDING_TIER_2` since we have no easy access to the device features and properties. A possible solution is to store a `vkd3d_physical_device_info` struct in `d3d12_device`.
- There is currently no clear path towards supporting bindless UAV counters. The original idea was to add a single SSBO descriptor to one of the bindless set layouts which would then contain the raw VAs, however this complicates the design significantly since we need to keep track of which binding to update etc. Another option which does not increase the overall set count would be to add it to the root descriptor set if necessary.
- We need a way to find the descriptor heap from a `D3D12_CPU_DESCRIPTOR_HANDLE`. The current (and somewhat ugly) solution is to initialize each descriptor with a pointer to the heap.
- There are still a few things I'd like to clean up, e.g. the way the descriptor table offset is calculated.

## Future work
- Inline uniform blocks should be used on implementations that only support 128 bytes of push constant space.
- Static samplers should be cached so that we don't exceed the limit on the number of samplers that may exist at a time. D3D12 allows up to [2032](https://docs.microsoft.com/en-us/windows/win32/direct3d12/hardware-support) unique static samplers.
- We may want to allocate static sampler descriptor sets from a global pool using `VK_DESCRIPTOR_POOL_CREATE_FREE_DESCRIPTOR_SET_BIT`. Currently we allocate a new set each time the root signature changes.
- Support volatile descriptor ranges properly for the packed descriptor set.
- Support dynamic descriptor indexing within the packed descriptor set.
- Use `VK_EXT_buffer_address` for root descriptors to replace the root descriptor set.
- Support bindless UAVs and UAV counters.
- Support bindless CBVs on hardware that can support it.

HansKristian-Work commented on include/private/vkd3d_common.h:
```diff
@@ -81,6 +82,36 @@ static inline bool vkd3d_bitmask_is_contiguous(unsigned int mask)
+static inline unsigned int vkd3d_bitmask_iter64(uint64_t* mask)
+{
+    uint64_t cur_mask = *mask;
+    *mask = cur_mask & (cur_mask - 1);
```
Is this correct? I would expect something like
```
*mask = ~(1ull << tzcnt(*mask));
```

doitsujin replied:
The idea is to clear the lowest set bit and leave the upper bits intact, which is what this does.

HansKristian-Work replied:
That's what I figured, but I have a hard time seeing how that mask does that, it is some classic "hackers delight" trick? I've never seen that way of doing a bitloop.

HansKristian-Work replied:
A comment explaining what the trick is doing is probably good enough.

HansKristian-Work commented on include/private/vkd3d_common.h:
```diff
@@ -81,6 +82,36 @@ static inline bool vkd3d_bitmask_is_contiguous(unsigned int mask)
+    unsigned long result;
+    _BitScanForward64(&result, mask) ? result : 64;
+    return result;
+#elif defined (HAVE_BUILTIN_CTZLL)
```
Is this activated in autoconf?

doitsujin replied:
Apparently I forgot to commit the change.

HansKristian-Work commented:
vkd3d: Rework descriptor table binding refers to a descriptor update template. I don't see that being used. Thought we dropped it?

HansKristian-Work commented on libs/vkd3d/command.c:
```diff
@@ -2538,176 +2544,186 @@ static bool d3d12_command_list_update_graphics_pipeline(struct d3d12_command_lis
+    unsigned int write_count = 0;
 
-    descriptor_table = root_signature_get_descriptor_table(root_signature, index);
+    if (!vkd3d_descriptor_updates_resize_arrays(updates, root_signature->packed_descriptor_count))
```
I suppose for RS 1.0 we can just append new updates here and flush on submit instead of updating immediately?

doitsujin replied:
Haven't thought about that, but yeah, that might actually work.

doitsujin replied:
Actually, it's not quite as simple and would need a (rather minor) rework of how we populate the array; it still shows that this was originally written with descriptor update templates in mind. But yes, I think this is the right way forward.

doitsujin replied:
Nevermind, doesn't work. We need to pull in the correct *descriptors* at submission time, so nothing really changes from your previous attempt at implementing it.

doitsujin commented:
> vkd3d: Rework descriptor table binding refers to a descriptor update template. I don't see that being used. Thought we dropped it?

I fixed the commit message.

HansKristian-Work commented on libs/vkd3d/vkd3d_private.h:
```diff
@@ -618,8 +623,32 @@ struct d3d12_descriptor_heap
+
+static inline unsigned int d3d12_descriptor_heap_cbv_set_index()
+{
+    return 0;
```
Is this because we haven't really bothered hooking up CBV bindless yet?

doitsujin replied:
These are intended to be somewhat hard-coded (but therefore, very fast) functions to give us access to a descriptor set stored with a descriptor heap. The exact numbers don't really matter. This wouldn't change with bindless CBV.

HansKristian-Work commented on libs/vkd3d/vkd3d_private.h:
```diff
@@ -618,8 +623,32 @@ struct d3d12_descriptor_heap
+
+static inline unsigned int d3d12_descriptor_heap_uav_set_index(bool is_buffer)
+{
+    return 3 + (is_buffer ? 0 : 1);
```
I suppose there is no argument for counters since we don't bother with bindless UAV counters yet?

doitsujin replied:
Kind of, in practice (assuming they get a separate set, or are incorporated into an existing set) we'd probably define a new function for UAV counters instead.

HansKristian-Work commented on libs/vkd3d/vkd3d_private.h:
```diff
@@ -647,6 +647,12 @@ static inline unsigned int d3d12_descriptor_heap_uav_set_index(bool is_buffer)
+static inline uint32_t d3d12_desc_heap_offset(const struct d3d12_desc *dst)
+{
+    size_t heap_start = (size_t)dst->heap->descriptors;
+    return ((size_t)dst - heap_start) / sizeof(struct d3d12_desc);
```
Is this sizeof aligned to POT? Integer divide isn't that cool. Alternatively, we can store the offset directly.

doitsujin replied:
> Is this sizeof aligned to POT?

No, but storing the offset (alongside the pointer to the heap, which we already have anyway) works too.

HansKristian-Work commented on libs/vkd3d/command.c:
```diff
@@ -2725,6 +2728,16 @@ static void d3d12_command_list_update_packed_descriptors(struct d3d12_command_li
     }
+
+    /* Set descriptor offsets */
+    if (root_signature->descriptor_table_count)
```
This is called only inside update_packed_descriptors. This seems a bit sketchy if we only changed a root descriptor which uses descriptor tables. I think we technically don't need to update packed descriptors in that case, only push constants.

doitsujin replied:
True, on the other hand we currently have no way to check whether a table has packed descriptors or not.

HansKristian-Work replied:
If any SetDescriptorTable call will invalidate the packed set, I suppose this is fine, but I'd like a TODO comment explaining what optimization we can do later so we don't forget we need to move this out.

doitsujin replied:
Added a comment.

I guess for an implementation, we could rework this further and not track each individual table, instead just adding a `packed_descriptors_dirty` and `descriptor_tables_dirty` flag similar to the static samplers one, and store whether a table has packed descriptors with the root signature.

HansKristian-Work replied:
ye, I think that'd be a good thing to do since it's unlikely UAV/CBV ranges will overlap with SRVs for bindless scenarios. But it doesn't block this PR.

HansKristian-Work commented on libs/vkd3d/device.c:
```diff
@@ -1350,12 +1353,20 @@ static HRESULT vkd3d_init_device_caps(struct d3d12_device *device,
+            && descriptor_indexing->descriptorBindingPartiallyBound
+            && descriptor_indexing->descriptorBindingVariableDescriptorCount
+            && descriptor_indexing_properties->maxPerStageDescriptorUpdateAfterBindSampledImages >= 1000000)
         device->feature_options.ResourceBindingTier = D3D12_RESOURCE_BINDING_TIER_2;
```
We'll probably need to report TIER_3 here anyways because WoW is broken and refuses to use D3D12 unless that is reported. Is there an "app-opt" thing in place we can use for this?

doitsujin replied:
> Is there an "app-opt" thing in place we can use for this?

No. I guess the best thing to do for now is to keep the old code for this and instead check for bindless features in `vkd3d_bindless_state_get_bindless_flags`.

HansKristian-Work replied:
That works, please add comments explaining the situation in that case.

doitsujin replied:
Done.

HansKristian-Work commented on libs/vkd3d/state.c:
```diff
@@ -343,42 +370,68 @@ struct d3d12_root_signature_info
+        {
+            case D3D12_DESCRIPTOR_RANGE_TYPE_SRV:
+                /* separate image + buffer descriptors */
+                info->binding_count += 2;
```
Probably best to set descriptor_count here and increment outside the branch. The two branches here to the same thing in different ways.

doitsujin replied:
The difference is that bindings and packed descriptors are decoupled here. Can of course introduce a separate `binding_count`, but this will lead to some duplication in the second branch.

HansKristian-Work replied:
Hm, right, nvm.

HansKristian-Work commented on libs/vkd3d/state.c:
```diff
@@ -712,9 +807,11 @@ static HRESULT d3d12_root_signature_init_static_samplers(struct d3d12_root_signa
         struct d3d12_device *device, const D3D12_ROOT_SIGNATURE_DESC *desc)
 {
+    const struct vkd3d_bindless_state *bindless_state = &device->bindless_state;
+    VkDescriptorSetLayout set_layouts[3 + VKD3D_MAX_BINDLESS_DESCRIPTOR_SETS];
```
If VKD3D_MAX_BINDLESS_DESCRIPTOR_SETS is 6, wouldn't this overflow the limit of 8?

doitsujin replied:
This is really just here so that we have a large enough array. I can hard-code it to 8 if you want.

We'll never have >8 sets in practice,. since if everything is bindless, there will be no packed descriptor set.

HansKristian-Work replied:
Defining VKD3D_MAX_DESCRIPTORS to 8 is probably fine.

doitsujin replied:
Done.

doitsujin commented on libs/vkd3d/state.c:
```diff
@@ -2764,8 +2764,22 @@ static HRESULT vkd3d_bindless_state_add_binding(struct vkd3d_bindless_state *bin
+            !device_info->descriptor_indexing_features.descriptorBindingVariableDescriptorCount)
+        return 0;
+
+    if (device_info->descriptor_indexing_properties.maxPerStageDescriptorUpdateAfterBindSampledImages >= 1000000 &&
```
Yeah, fixed.
